drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+			dev_data->rx_errors = 0;
+			dev_data->tx_errors = 0;
+		}
+
+		/* Execute state change callback if registered */
+		if (dev_data->state_chg_cb != NULL) {
+			LOG_DBG("%s calling state change callback function",
+				dev->name);
+
+			error_counters.rx_err_cnt = dev_data->rx_errors;
+			error_counters.tx_err_cnt = dev_data->tx_errors;
+			dev_data->state_chg_cb(dev, state, error_counters,
+					       dev_data->state_chg_cb_data);
+		}
+	}
+}
+
+/**
+ * @brief Bus error handling function
+ * @internal
+ *
+ * Called from within the ISR if the error interrupt bit is set in
+ * the interrupt status register, this function evaluates the Error
+ * Status Register (ESR) and translates the register's contents into
+ * detailed error counters. Subsequently, the Error Counter Register
+ * (ECR) is read and the RX error or TX error counter (whichever is
+ * greater) is used to determine an updated CAN bus state.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_handle_errors(const struct device *dev)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+	struct can_xlnx_dev_data *dev_data = dev->data;
+	uint32_t esr;
+	uint32_t ecr;
+	uint8_t higher_errors;
+
+	/* Decode the type of error -> update error statistics */
+	esr = sys_read32(base_addr + CAN_XLNX_ESR_OFFSET);
+	LOG_ERR("%s error(s): %s%s%s%s%s\n",
+		dev->name,
+		(esr & CAN_XLNX_ESR_ACK_ERROR)   ? "ACK " : "",
+		(esr & CAN_XLNX_ESR_BIT_ERROR)   ? "BIT " : "",
+		(esr & CAN_XLNX_ESR_STUFF_ERROR) ? "STF " : "",
+		(esr & CAN_XLNX_ESR_FORM_ERROR)  ? "FRM " : "",
+		(esr & CAN_XLNX_ESR_CRC_ERROR)   ? "CRC " : "");
+
+	if (esr & CAN_XLNX_ESR_ACK_ERROR) {
+		CAN_STATS_ACK_ERROR_INC(dev);
+	}
+	if (esr & CAN_XLNX_ESR_BIT_ERROR) {
+		CAN_STATS_BIT0_ERROR_INC(dev);
+	}
+	if (esr & CAN_XLNX_ESR_STUFF_ERROR) {
+		CAN_STATS_STUFF_ERROR_INC(dev);
+	}
+	if (esr & CAN_XLNX_ESR_FORM_ERROR) {
+		CAN_STATS_FORM_ERROR_INC(dev);
+	}
+	if (esr & CAN_XLNX_ESR_CRC_ERROR) {
+		CAN_STATS_CRC_ERROR_INC(dev);
+	}
+
+	/* Clear the error status register */
+	sys_write32(CAN_XLNX_ESR_CLEAR_ALL_MASK, base_addr + CAN_XLNX_ESR_OFFSET);
+
+	/* Update the error counters */
+	ecr = sys_read32(base_addr + CAN_XLNX_ECR_OFFSET);
+	dev_data->rx_errors = (uint8_t)((ecr >> CAN_XLNX_ECR_RX_ERRORS_OFFSET) &
+		CAN_XLNX_ECR_RX_ERRORS_MASK);
+	dev_data->tx_errors = (uint8_t)((ecr >> CAN_XLNX_ECR_TX_ERRORS_OFFSET) &
+		CAN_XLNX_ECR_TX_ERRORS_MASK);
+
+	/*
+	 * Check if a state change is required due to the new error counter values.
+	 * Bus off state can not be detected via the error counters, entering this
+	 * error state is indicated by a separate interrupt flag.
+	 *
+	 * NOTICE: this could alternatively be determined by evaluating the ESTAT
+	 * bits in the Status Register (SR). Comp. Zynq-7000 TRM, register SR de-
+	 * scription, Appendix B.5, p. 804), but we need the counters updated any-
+	 * ways as we need up-to-date values whenever the counters are queried
+	 * via can_xlnx_get_state.
```

Comment:
Or you could just read out the counter values in `can_xlnx_get_state()`. I would prefer determining the error state using the state itself and not the error counters.